add ability to download from tar.gz; add tests

rwalle61 opened the pull request:
Signed-off-by: Richard Waller <Richard.Waller@ibm.com>

## Problem
- The work done under https://github.com/eclipse/codewind/pull/62 has added the ability to create projects from urls that point to .tar.gz files. We need to add the same functionality to the installer.
- Our existing download functions were not covered by tests.

## Solution
- Write a function to download and extract `.tar.gz` files from appropriate URLs
- Cover our download functions in tests.

## Tested
Automatically. Output: 
<img width="1134" alt="image" src="https://user-images.githubusercontent.com/18170169/63039448-35115180-bebb-11e9-93d8-3278b87595c0.png">

## Extra
- Renamed some variables for clarity (e.g. `tempFilePath` was actually a generic `filePath`)
- Make functions return errors so that we can handle them appropriately

liamchampton commented on actions/project.go:
```diff
@@ -49,43 +49,80 @@ func DownloadTemplate(c *cli.Context) {
+	url := c.String("u")
 
+	err := DownloadFromURLThenExtract(url, destination)
+	if err != nil {
```
Can this use the error checking module? - was created to avoid having the code littered with this `if statement`

liamchampton commented on actions/project.go:
```diff
@@ -49,43 +49,80 @@ func DownloadTemplate(c *cli.Context) {
+
+	pathToTempFile := destination + "/temp.tar.gz"
+	err := utils.DownloadFile(URL, pathToTempFile)
+	if err != nil {
```
Same as above, can this be moved?

liamchampton commented on actions/project.go:
```diff
@@ -49,43 +49,80 @@ func DownloadTemplate(c *cli.Context) {
-	} else {
-		tempPath = "/tmp/"
+	zipURL, err := utils.GetZipURL(owner, repo, branch)
+	if err != nil {
```
Same as above

liamchampton commented on actions/project.go:
```diff
@@ -49,43 +49,80 @@ func DownloadTemplate(c *cli.Context) {
-	if err := utils.DownloadFile(zipFileName, zipURL); err != nil {
-		log.Fatal(err)
+	err := utils.DownloadFile(zipURL, pathToTempZipFile)
+	if err != nil {
```
Same as above

liamchampton commented on actions/project.go:
```diff
@@ -49,43 +49,80 @@ func DownloadTemplate(c *cli.Context) {
-	// unzip into /tmp dir
-	utils.UnZip(zipFileName, destination)
+	err = utils.UnZip(pathToTempZipFile, destination)
+	if err != nil {
```
Same as above

liamchampton commented on utils/filesystem.go:
```diff
@@ -72,17 +73,16 @@ func WriteToComposeFile(tempFilePath string, debug bool) bool {
-	os.Remove(tempFilePath)
-	fmt.Println("==> finished deleting file " + tempFilePath)
+	os.Remove(filePath)
+	log.Printf("Deleted file: %s\n\n", filePath)
```
What is the reason for changing this to `log`? By default `log` prints to stderr. I dont think this needs to be in stderr and `fmt` is sufficient.

liamchampton commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
-	resp, err := http.Get(url)
-	errors.CheckErr(err, 400, "")
+	resp, err := http.Get(URL)
+	if err != nil {
```
same comment as above, can this not be moved into the error module?

liamchampton commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
-	file, err := os.Create(zipFileName)
-	errors.CheckErr(err, 401, "")
+	file, err := os.Create(destination)
+	if err != nil {
```
same as above

liamchampton commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
 	// Write body to file
 	_, err = io.Copy(file, resp.Body)
-	fmt.Println(zipFileName)
+	log.Printf("Downloaded file: from '%s' to '%s'\n", URL, destination)
```
Again, i dont think this is necessary to change `fmt` to `log`. We dont need to print to stderr here. `fmt.Printf()` should be enough here.

liamchampton commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
 
 		if file.FileInfo().IsDir() {
-			log.Println("Directory Created:", extractedFilePath)
+			// log.Println("Directory Created:", extractedFilePath)
```
whats the reason for this comment?

liamchampton commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
 			os.MkdirAll(extractedFilePath, file.Mode())
 		} else {
-			log.Println("File extracted:", file.Name)
+			// log.Println("File extracted:", file.Name)
```
and this comment?

liamchampton commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+// UnTar unpacks a tar.gz file to a destination
+func UnTar(pathToTarFile, destination string) error {
+	fileReader, err := read(pathToTarFile)
+	if err != nil {
```
Can this be moved to use the module?

liamchampton commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+	}
+	defer fileReader.Close()
+	gzipReader, err := gzip.NewReader(fileReader)
+	if err != nil {
```
same as above

liamchampton commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+			break
+		}
+		if err != nil {
+			panic(err)
```
Is there a reason for using a `panic` here? This will print the entire stack trace (including build path) which is very verbose. I think `log.fatal` is enough information to print out and will automatically be followed by `os.Exit(1)`. - this could also be moved to the error module

rwalle61 commented:
Thanks for the review. I've corrected some `log` to `fmt`, switched `panic` to `log.Fatal`, and moved my additions to `project.go` to `utils/download.go`. 

As we discussed, we will keep my error handling for now, and next look to synthesise it with the existing approach to make a better one.

jcockbain96 commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+}
+
+func overwrite(filePath string) (*os.File, error) {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_TRUNC, 0777)
```
Comment permissions here, as with others?

jcockbain96 commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+}
+
+func read(filePath string) (*os.File, error) {
+	file, err := os.OpenFile(filePath, os.O_RDONLY, 0444)
```
Could also comment permissions here if you think it's needed?

jcockbain96 commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+	return file, nil
+}
+
+func read(filePath string) (*os.File, error) {
```
It's probably okay because the parameter is filePath (and it's only local) but could consider a more descriptive name for this function? Would you also want to export this though in future?

jcockbain96 commented on utils/download.go:
```diff
@@ -0,0 +1,82 @@
+	return nil
+}
+
+
```
Double blank line here?

jcockbain96 commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
-//UnZip downloaded file
-func UnZip(zipFileName, fileDestination string) {
-	zipReader, _ := zip.OpenReader(zipFileName)
+//UnZip unzips a file to a destination
```
Looking back is actually in loads of places. defo a nit but prefer with a space

jcockbain96 commented on utils/download.go:
```diff
@@ -0,0 +1,82 @@
+func DownloadFromTarGzURL(URL string, destination string) error {
+	_ = os.MkdirAll(destination, 0700) // gives User rwx permission
+
+	pathToTempFile := destination + "/temp.tar.gz"
```
or filepath.Join or whatever?

rwalle61 replied:
good spot, I meant to change that

rwalle61 commented on utils/filesystem.go:
```diff
@@ -72,17 +73,16 @@ func WriteToComposeFile(tempFilePath string, debug bool) bool {
-	os.Remove(tempFilePath)
-	fmt.Println("==> finished deleting file " + tempFilePath)
+	os.Remove(filePath)
+	// fmt.Printf("==> Deleted file: %s\n", filePath)
```
for debugging

rwalle61 commented on utils/filesystem.go:
```diff
@@ -113,63 +113,69 @@ func PingHealth(healthEndpoint string) bool {
 
 		if file.FileInfo().IsDir() {
-			log.Println("Directory Created:", extractedFilePath)
+			// fmt.Println("Directory Created:", extractedFilePath)
```
for debugging. I suppose we could improve our logging system to have levels

jcockbain96 replied:
I assumed so. Not a big deal. Could make that clear with the comment but don't mind

liamchampton commented on utils/filesystem.go:
```diff
@@ -183,5 +189,69 @@ func UnZip(zipFileName, fileDestination string) {
+				log.Fatal(err)
+			}
+		default:
+			log.Printf("Can't: %c, %s\n", header.Typeflag, target)
```
does this need to go to stderr? Also unsure what this is defaulting too. Maybe a comment?

rwalle61 replied:
I think it does need to go to stderr because it's a failure (while unpacking a tar.gz). I've clarified the log message

liamchampton commented on utils/download_test.go:
```diff
@@ -0,0 +1,231 @@
+import (
+	"errors"
+	"io/ioutil"
+
```
This new line isnt needed. Removing it will keep all internal library imports together
